### DISCUSSION TITLE
How to use MCP23017 Bank as input

### DISCUSSION THREAD

**Discussion by orrious** [tagyoureit/nodejs-poolController#529] [REM]
In REM, how do you define a MCP23017 Bank as input?  How do you set the GPIO pin on the RPi for that bank's interrupt?

**Comment by rstrouse** [comment]
I did not code any input functions for the I/O extenders.  The primary focus has been on extending the outputs based upon available relay boards at the time.  I haven't messed with the interrupts on that chip but I suspect that it operates like most I2c controllers with an interrupt where you monitor an external pin that is controlled at the device hardware layer (exported GPIO wired to the INT pin).   

An I2c slave cannot pull the line to send data without being asked (polled).  So the INT detection is via a pin and on the chip and in the MCP23017 there are two such pins.  I don't know how they need to be wired so they may need a transistor in between to monitor them but other I2c implementations that I have seen simply flop between 0v and Vcc.  The INTCON, GPINTEN, and DEFVAL registers determine how the interrupt is fired and the INTCAP registers reflect the state of the time of firing.

**Reply by orrious** [reply]
On the MCP23017 each bank has a INT (interrupt) pin that needs to be connected to a RPi GPIO.  My understanding is REM would need to monitor the GPIO, when it goes HI, go read the bank that is associated with it on the MCP23017.  After the bank has been read, the INT will go low until the next event.

You do not need a transistor or level shifter if you are running the MCP23017 at 3.3v.  If you run the MCP23017 at 5v, I would use a level shifter or voltage divider to drop the voltage between INT and RPi GPIO.
